File: epi_judge_java/epi/KthLargestElementInLongArray.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.Iterator;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Queue;

public class KthLargestElementInLongArray {

    public static int findKthLargestUnknownLength(Iterator<Integer> stream, int k) {
        Queue<Integer> minHeap = new PriorityQueue<>();
        for (int i = 0; i < k; i++) {
            minHeap.offer(stream.next());
        }

        while (stream.hasNext()) {
            minHeap.offer(stream.next());
            minHeap.poll();
        }

        return minHeap.poll();
    }

    @EpiTest(testDataFile = "kth_largest_element_in_long_array.tsv")
    public static int findKthLargestUnknownLengthWrapper(List<Integer> stream,
                                                         int k) {
        return findKthLargestUnknownLength(stream.iterator(), k);
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "KthLargestElementInLongArray.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/kth_largest_element_in_long_array.cc
#include <vector>

#include "test_framework/generic_test.h"
using namespace std;

int FindKthLargestUnknownLength(vector<int>::const_iterator stream_begin,
                                const vector<int>::const_iterator &stream_end,
                                int k) {

    priority_queue<int, vector<int>, greater<>> min_heap;
    auto it = stream_begin;
    while (it != stream_end) {
        min_heap.push(*it);
        if (min_heap.size() > k) {
            min_heap.pop();
        }
        ++it;
    }

    return min_heap.top();
}

int FindKthLargestUnknownLengthWrapper(const vector<int> &stream, int k) {
    return FindKthLargestUnknownLength(stream.cbegin(), stream.cend(), k);
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"stream", "k"};
    return GenericTestMain(args, "kth_largest_element_in_long_array.cc",
                           "kth_largest_element_in_long_array.tsv",
                           &FindKthLargestUnknownLengthWrapper,
                           DefaultComparator{}, param_names);
}
